ironfish-rust/src/public_account/description.rs

```diff
@@ -0,0 +1,218 @@
+            let signature = Signature::from_bytes(&signature_buf);
+            signatures.push(signature);
+        }
+
+        let mut transfers_len_buf = [0; 2];
+        reader.read_exact(&mut transfers_len_buf)?;
+        let transfers_len = i16::from_le_bytes(transfers_len_buf) as usize;
+
+        let mut transfers = Vec::with_capacity(transfers_len);
+        for _ in 0..transfers_len {
+            let transfer = Transfer::read(&mut reader)?;
+            transfers.push(transfer);
+        }
+        let address = PublicAddress::read(&mut reader)?;
+
+
+        Ok(Self {
+            version,
+            threshold,
+            signers,
+            signatures,
+            address,
+            transfers,
+        })
+    }
+
+    pub fn write<W: io::Write>(&self, mut writer: W) -> Result<(), IronfishError> {
+        // TODO: think about ideal ordering here
+        writer.write_all(&self.version.to_le_bytes())?;
+        
+        writer.write_all(&self.threshold.to_le_bytes())?;
+
+        let signers_len = self.signers.len() as i16;
+        writer.write_all(&signers_len.to_le_bytes())?;
+
+        for signer in &self.signers {
+            writer.write_all(signer.as_bytes())?;
+        }
+
+        let signatures_len = self.signatures.len() as i16;
+        writer.write_all(&signatures_len.to_le_bytes())?;
+
+        for signature in &self.signatures {
+            writer.write_all(&signature.to_bytes())?;
+        }
+
+        let transfers_len = self.transfers.len() as i16;
+        writer.write_all(&transfers_len.to_le_bytes())?;
+
+        for transfer in &self.transfers {
+            transfer.write(&mut writer)?;
+        }
+
+        self.address.write(&mut writer)?;
+
+        Ok(())
+    }
+
+    pub fn valid(&self) -> Result<(), IronfishError> {
+        if self.threshold < 1 {
+            return Err(IronfishError::new(IronfishErrorKind::InvalidThreshold));
+        }
+
+        if self.signers.len() < self.threshold as usize {
+            return Err(IronfishError::new(IronfishErrorKind::InvalidThreshold));
+        }
+
+        if self.signatures.len() < self.threshold as usize {
+            return Err(IronfishError::new(IronfishErrorKind::InvalidData));
+        }
+
+        Ok(())
+    }
+
+    pub fn verify(&self) -> Result<(), IronfishError> {
+        self.valid()?;
+
+        let hash =
+            &PublicAccountDescription::hash(&self.version, &self.threshold, &self.signers, &self.address, &self.transfers)?;
+        let is_valid = self
+            .signers
+            .iter()
+            .zip(&self.signatures)
+            .any(|(signer, signature)| signer.verify(hash, signature).is_ok());
+        if !is_valid {
+            return Err(IronfishError::new(IronfishErrorKind::InvalidSignature));
+        }
+        Ok(())
+    }
+
+    pub fn hash(
+        version: &u8,
+        threshold: &i16,
+        signers: &Vec<VerifyingKey>,
+        address: &PublicAddress,
+        transfers: &Vec<Transfer>,
```

Comment:
Why not make this a `&self`?

Reply:
Sg, I can do that as long as I instantiate with an empty vec for signatures (instead of requiring them to be present on instantiation). I updated to do that.